Failed to build `ghdl_llvm` on Linux (x86_64):
```
./ortho_code-x86-abi.o: In function `ortho_code__x86__abi__link_intrinsics':
/tmp/rasen/nix-build-ghdl-mcode-0.33.drv-0/ghdl-v0.33-src/src/ortho/mcode/ortho_code-x86-abi.adb:775: undefined reference to `__muldi3'
/tmp/rasen/nix-build-ghdl-mcode-0.33.drv-0/ghdl-v0.33-src/src/ortho/mcode/ortho_code-x86-abi.adb:779: undefined reference to `__divdi3'
/tmp/rasen/nix-build-ghdl-mcode-0.33.drv-0/ghdl-v0.33-src/src/ortho/mcode/ortho_code-x86-abi.adb:783: undefined reference to `__moddi3'
collect2: ld returned 1 exit status
gnatlink: error when calling /nix/store/6qqkwjbga7gcsywl028zw2pvhy43czr2-gnat-wrapper-4.5.4/bin/gcc
gnatmake: *** link failed.
make: *** [Makefile:95: ghdl_mcode] Error 4
builder for ‘/nix/store/vd5k2fzszak1mb89bqy5ysw0kdwhwglw-ghdl-mcode-0.33.drv’ failed with exit code 2
error: build of ‘/nix/store/vd5k2fzszak1mb89bqy5ysw0kdwhwglw-ghdl-mcode-0.33.drv’ failed
```

--- a/pkgs/development/compilers/ghdl/default.nix
+++ b/pkgs/development/compilers/ghdl/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchurl, gnat, zlib, llvm_35, ncurses, clang, flavour ? "mcode" }:
+{ stdenv, fetchFromGitHub, gnat, zlib, llvm_35, ncurses, clang, flavour ? "mcode" }:
 
 # mcode only works on x86, while the llvm flavour works on both x86 and x86_64.
 
@@ -13,9 +13,11 @@ in
 stdenv.mkDerivation rec {
   name = "ghdl-${flavour}-${version}";
 
-  src = fetchurl {
-    url = "https://github.com/tgingold/ghdl/archive/v${version}.tar.gz";
-    sha256 = "09yvgqyglbakd74v2dgr470clzm744i232nixyffcds55vkij5da";
+  src = fetchFromGitHub {
+    owner = "tgingold";
+    repo = "ghdl";
+    rev = "v${version}";
+    sha256 = "0g72rk2yzr0lrpncq2c1qcv71w3mi2hjq84r1yzgjr6d0qm87r2a";
   };
 
   buildInputs = [ gnat zlib ] ++ optionals (flavour == "llvm") [ clang ncurses ];

--- a/pkgs/development/compilers/ghdl/default.nix
+++ b/pkgs/development/compilers/ghdl/default.nix
@@ -1,16 +1,26 @@
-{ stdenv, fetchurl, gnat, zlib }:
+{ stdenv, fetchurl, gnat, zlib, llvm_35, ncurses, clang, flavour ? "mcode" }:
+
+# mcode only works on x86, while the llvm flavour works on both x86 and x86_64.
+
+
+assert flavour == "llvm" || flavour == "mcode";
+
 let
+  inherit (stdenv.lib) optional;
+  inherit (stdenv.lib) optionals;
   version = "0.33";
 in
 stdenv.mkDerivation rec {
-  name = "ghdl-mcode-${version}";
+  name = "ghdl-${flavour}-${version}";
 
   src = fetchurl {
     url = "https://github.com/tgingold/ghdl/archive/v${version}.tar.gz";
     sha256 = "09yvgqyglbakd74v2dgr470clzm744i232nixyffcds55vkij5da";
   };
 
-  buildInputs = [ gnat zlib ];
+  buildInputs = [ gnat zlib ] ++ optionals (flavour == "llvm") [ clang ncurses ];
+
+  configureFlags = optional (flavour == "llvm") "--with-llvm=${llvm_35}";
 
   patchPhase = ''
     # Disable warnings-as-errors, because there are warnings (unused things)
@@ -23,11 +33,10 @@ stdenv.mkDerivation rec {
 
   meta = {
     homepage = "http://sourceforge.net/p/ghdl-updates/wiki/Home/";
-    description = "Free VHDL simulator, mcode flavour";
+    description = "Free VHDL simulator";
     maintainers = with stdenv.lib.maintainers; [viric];
-    # I think that mcode can only generate x86 code,
-    # so it fails to link pieces on x86_64.
-    platforms = with stdenv.lib.platforms; [ "i686-linux" ];
+    platforms = with stdenv.lib.platforms; (if flavour == "llvm" then [ "i686-linux" "x86_64-linux" ]
+      else [ "i686-linux" ]);
     license = stdenv.lib.licenses.gpl2Plus;
   };
 }

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -4710,7 +4710,13 @@ in
     profiledCompiler = false;
   });
 
-  ghdl_mcode = callPackage_i686 ../development/compilers/ghdl { };
+  ghdl_mcode = callPackage_i686 ../development/compilers/ghdl {
+    flavour = "mcode";
+  };
+
+  ghdl_llvm = callPackage ../development/compilers/ghdl {
+    flavour = "llvm";
+  };
 
   gcl = callPackage ../development/compilers/gcl {
     gmp = gmp4;

